common/fs/src/cache/tailed_file.rs

```diff
@@ -0,0 +1,109 @@
+use http::types::body::LineBuilder;
+use metrics::Metrics;
+
+use futures::Stream;
+
+use std::fs::{File, OpenOptions};
+use std::io::{BufRead, BufReader, Seek, SeekFrom};
+use std::path::{Path, PathBuf};
+
+#[derive(Debug)]
+pub struct TailedFile {
+    reader: std::io::BufReader<File>,
+    offset: u64,
+}
+
+impl TailedFile {
+    pub(crate) fn new(path: &Path) -> Result<Self, std::io::Error> {
+        Ok(Self {
+            reader: BufReader::new(OpenOptions::new().read(true).open(path)?),
+            offset: 0,
+        })
+    }
+
+    pub(crate) fn seek(&mut self, offset: u64) -> Result<(), std::io::Error> {
+        if let Err(e) = self.reader.seek(SeekFrom::Start(offset)) {
+            error!("error seeking {:?}", e);
+            return Err(e);
+        }
+        self.offset = offset;
+        Ok(())
+    }
+    // tail a file for new line(s)
+    pub(crate) fn tail(&mut self, paths: &[PathBuf]) -> Option<impl Stream<Item = LineBuilder>> {
+        // get the file len
+        let len = match self.reader.get_ref().metadata().map(|m| m.len()) {
+            Ok(v) => v,
+            Err(e) => {
+                error!("unable to stat {:?}: {:?}", &paths[0], e);
+                return None;
+            }
+        };
+
+        // if we are at the end of the file there's no work to do
+        if self.offset == len {
+            return None;
+        }
+
+        // if the offset is greater than the file's len
+        // it's very likely a truncation occurred
+        if self.offset > len {
+            info!(
+                "{:?} was truncated from {} to {}",
+                &paths[0], self.offset, len
+            );
+            // Reset offset back to the start... ish?
+            // TODO: Work out the purpose of the 8192 something to do with lookback? That seems wrong.
```

Comment:
It could be something to do with a truncate logrotation strategy trancating to 8k or something? I think it's worth digging into separately in case removing it makes us start duplicating 8k or logs every time a log file is rotated.

We should really work out a better system here thought...